### PR TITLE
microservice-architectures

### PR DESCRIPTION
**Pull request by temyers**

### CODE REVIEW

File: _posts/2020-05-18-microservice-architectures.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+Events comprise of two core elements:
+* Domain event - the important detail of the event and includes important attributes such as timestamp, and eventType.  We use the [cloud event specification](https://cloudevents.io/) for defining domain events.
+* Envelope - wraps the domain event and captures details of the event transport, such as [SNS](https://docs.aws.amazon.com/sns/latest/dg/sns-message-and-json-formats.html) or [EventBridge](https://docs.aws.amazon.com/eventbridge/latest/userguide/aws-events.html) events.
+
+![clean architecture diagram showing onion of layers including handler, integration with ports and adapters, service and domain layers ](/img/microservice-architectures/clean-architecture.png)
+
+By following [clean architecture](https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html) principles, separating envelope unmarshalling from acting on domain events, provides additional flexibility on your deployment architecture.  For example, you can use the same domain processing logic to handle synchronous requests sent via API gateway or asynchronously via SNS.
```
Comment (matt-tyler) [inline _posts/2020-05-18-microservice-architectures.md:34]: _Following clean architecture principles (such as separating envelope unmarshalling from acting on domain events) provides additional flexibility in your deployment architecture._

??

File: _posts/2020-05-18-microservice-architectures.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+Irrespective of whether you use [CQRS](https://martinfowler.com/bliki/CQRS.html), events come in two categories:
+* commands 
+* events
+
+Commands are a request to do something: `requestCheckout`, `updateBalance`, `createMyWidget`.  Commands are inherently transient, may be resubmitted multiple times and may/may not result in action being taken.  Events are a notification that something has happened, or a particular state: `checkoutCompleted`, `balanceUpdated`, `MyWidgetCreated`.  Events occurred at a point in in the past.  Depending on how you choose to interpret events, a key part of the semantics, event re-delivery may be possible.  For example, `accountCreated` may be interpreted as "an account has _just_ been created", or "an account was created at a certain point in time".  The former has implications on event re-delivery, whereas the latter is immutable and less affected by the temporal difference between when the actual creation occurred and when the event was generated.
```
Comment (matt-tyler) [inline _posts/2020-05-18-microservice-architectures.md:40]: _Commands are inherently transient, may be resubmitted multiple times, and may/may not result in action being taken._

I feel like you need an oxford comma here.

Comment (matt-tyler) [inline _posts/2020-05-18-microservice-architectures.md:40]: _a key part of the semantics_

I feel you can lose this phrase and not lose the meaning of the sentence.

File: _posts/2020-05-18-microservice-architectures.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+![rivers rapids pattern showing a single event bus: rapids; topic specific buses: rivers; and microservice consumers: ponds](/img/microservice-architectures/rivers-rapids.png)
+
+## Domain Oriented
+
+Event driven systems complement, and necessitate, [domain driven design](https://martinfowler.com/tags/domain%20driven%20design.html).  The focus on domain events identify natural service boundaries and decouple producers and consumers.  As a result, a microservice often comprises of a collection of highly cohesive, tightly coupled functions.  Using event boundaries, we are able to disregard the inherent complexity in the production of events the service consumes.  The generation of a [`CreateManagedAccount`](https://docs.aws.amazon.com/controltower/latest/userguide/lifecycle-events.html#create-managed-account) contains a large amount of complexity, orchestrating the creation of a new AWS account.  If you want to deploy a Config Rule, as part of your cloud governance strategy, to all new accounts you can ignore all that incidental complexity and simply respond to the event.
```
Comment (matt-tyler) [inline _posts/2020-05-18-microservice-architectures.md:56]: _Event driven systems complement and necessitate domain driven design._

Can lose the commas; the sentence is short enough.

Comment (matt-tyler) [reply _posts/2020-05-18-microservice-architectures.md:56]: _The generation of a [`CreateManagedAccount`](https://docs.aws.amazon.com/controltower/latest/userguide/lifecycle-events.html#create-managed-account) *event abstracts the complexity of orchestrating the creation of a new AWS account*_

_If you want to deploy a Config Rule as part of your cloud governance strategy to all new accounts, you can ignore all that incidental complexity and simply respond to the event._

File: _posts/2020-05-18-microservice-architectures.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+date: 2020-05-05
+tags: cloud microservices architecture
+author: Matthew Tyler
+image: img/microservice-architectures/death-star.jpg
+---
+
+<center><img src="/img/microservice-architectures/death-star.jpg" /></center>
+<br/>
+
+At Mechanical Rock, we build [Cloud Native solutions](https://www.mechanicalrock.io/our-expertise/cloud-native-solutions/), in [small teams](https://www.inc.com/business-insider/jeff-bezos-productivity-tip-two-pizza-rule.html).  We find building using microservice patterns highly effective to simplify our overall approach.  Microservices help us to reduce the [cognitive load](https://youtu.be/haejb5rzKsM) when dealing with complex systems.  Furthermore, serverless technologies provide a natural fit for building microservice architectures that are not only decoupled and highly scalable, but also easy to reason about.
```
Comment (Hamo69) [inline _posts/2020-05-18-microservice-architectures.md:13]: > approach.  Microservices

double space madness

File: _posts/2020-05-18-microservice-architectures.md (hunk 3 above)
Comment (Hamo69) [inline _posts/2020-05-18-microservice-architectures.md:56]: > and necessitate, [domain driven design]

unnecessary comma?

Comment (temyers) [reply _posts/2020-05-18-microservice-architectures.md:56]: not in my opinion.

The intent is that it is a clarifying, yet incidental, detail.

**Pull request merged**